mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -0,0 +1,633 @@
+      inputAggBufferOffset: Int)
+    extends TypedImperativeAggregate[SummarizerBuffer] {
+
+    override def eval(state: SummarizerBuffer): InternalRow = {
+      val metrics = requestedMetrics.map({
+        case Mean => UnsafeArrayData.fromPrimitiveArray(state.mean.toArray)
+        case Variance => UnsafeArrayData.fromPrimitiveArray(state.variance.toArray)
+        case Count => state.count
+        case NumNonZeros => UnsafeArrayData.fromPrimitiveArray(
+          state.numNonzeros.toArray.map(_.toLong))
+        case Max => UnsafeArrayData.fromPrimitiveArray(state.max.toArray)
+        case Min => UnsafeArrayData.fromPrimitiveArray(state.min.toArray)
+        case NormL2 => UnsafeArrayData.fromPrimitiveArray(state.normL2.toArray)
+        case NormL1 => UnsafeArrayData.fromPrimitiveArray(state.normL1.toArray)
+      })
+      InternalRow.apply(metrics: _*)
+    }
+
+    override def children: Seq[Expression] = featuresExpr :: weightExpr :: Nil
+
+    override def update(state: SummarizerBuffer, row: InternalRow): SummarizerBuffer = {
+      // val features = udt.deserialize(featuresExpr.eval(row))
+      val featuresDatum = featuresExpr.eval(row).asInstanceOf[InternalRow]
+
+      val isDense = featuresDatum.getByte(0) == 1
+
+      val indices: ArrayData = if (isDense) null else featuresDatum.getArray(2)
+      val values: ArrayData = featuresDatum.getArray(3)
+      val _size = if (isDense) values.numElements() else featuresDatum.getInt(1)
+
+      val features = new TraversableIndexedSeq {
+
+        override def foreachActive(f: (Int, Double) => Unit): Unit = {
+          var i = 0
+          if (isDense) {
+            while (i < _size) {
+              f(i, values.getDouble(i))
+              i += 1
+            }
+          } else {
+            while (i < indices.numElements()) {
+              f(indices.getInt(i), values.getDouble(i))
+              i += 1
+            }
+          }
+        }
+
+        override def size: Int = _size
+      }
+
+      val weight = weightExpr.eval(row).asInstanceOf[Double]
+
+      state.addRaw(features, weight)
+      state
+    }
+
+    override def merge(state: SummarizerBuffer,
+                       other: SummarizerBuffer): SummarizerBuffer = {
+      state.merge(other)
+    }
+
+
+    override def nullable: Boolean = false
+
+    override def createAggregationBuffer(): SummarizerBuffer
+      = new SummarizerBuffer(requestedMetrics, requestedComputeMetrics)
+
+    override def serialize(state: SummarizerBuffer): Array[Byte] = {
+      // TODO: Use ByteBuffer to optimize
+      val bos = new ByteArrayOutputStream()
+      val oos = new ObjectOutputStream(bos)
+      oos.writeObject(state)
+      bos.toByteArray
+    }
+
+    override def deserialize(bytes: Array[Byte]): SummarizerBuffer = {
+      // TODO: Use ByteBuffer to optimize
+      val bis = new ByteArrayInputStream(bytes)
+      val ois = new ObjectInputStream(bis)
+      ois.readObject().asInstanceOf[SummarizerBuffer]
+    }
+
```

Comment:
I will optimize `serialize`/`deserialize` by `ByteBuffer` later. Though it is not the bottleneck currently.